### PR TITLE
Move native code into posixlib and clib respectively

### PR DESCRIPTION
**Pull request by ekrich**
The overall goal here is to move the C code into the corresponding library. We want to consider supporting Windows so allowing POSIX to be only available on UNIX like platforms makes sense.
 
This is just a first cut. We need to add `pthread.c` and a little more possibly that does not belong in the `nativelib`. Also, this will need to be repeated for `clib`. Most of the POSIX code has a `xxx.scala` file and a corresponding `xxx.c` file. `clib` was coded earlier before better conventions came along.

There is a question whether `posixlib` and the `nativelib` should `exportJars := true` in sbt. Removing this will still work fine and maybe build a little faster but to use the `SNAPSHOT` locally, `publishLocal` would have to be run. Let me know what you think about this.

### CODE REVIEW

**Comment by LeeTibbert** [comment]
Eric,

Thank you for stepping into this swamp and encouraging a healthy ecosystem.

Did you find a good decision rule for what should be in posixlib and what in clib.  
The reason that I ask is that I seem to remember doing a PR which later got 
overwhelmed by events which tried to move posixlib C files into a subdirectory
of resources. I think the intent was that once those had been coralled then the
clib C files into another: an early & aborted attempt similar to what you describe
here.

I seem to remember that I could find no clear separation of the two. That is
posix and c standards have become co-mingled over the years. IIRC, glibc
tends to implement  posix routines and a more. Recent posix standards tend
to refer to the corresponding IEEE C standard (and describe what to do if
there are conflicts).

I ran aground because I did not want to duplicate implementations and could not find a way
in GitHub to use links.

I think the same chaotic boundary exists in the .rst documentation.

It would be nice to have (read: IMHO, essential) to have a clearly stated
decision rule to guide both SN  and application developers.  

That is, which library is more primitive and can the less primitive be implemented in terms
of the primitive?    I believe clib is intended to be the primitive. Current posixlib impementations
may not be done in terms of the (intended underlying?) clib.

Pardon me if such exists and I missed it or was not clever enough to suss it out.

**Comment by LeeTibbert** [comment]
I believe there are at least 4 PRs in the queue which reference posixlib and which might
be affected by this PR.  PR #1635 is especially complicated and provides a major capability.
It requires two PRs which are awaiting merge, both of which may be affected.

Not to hold up the base PR, but I think it would ease the overall effort if those earlier PRs
were considered before this one.  Time is what keeps everything from happening at once.

**Comment by ekrich** [comment]
There are basically three parts to the situation.

1. There are the C routines that are needed in `nativelib` so that it doesn't depend on `clib` and visa versa. That is here.
 https://github.com/scala-native/scala-native/blob/master/nativelib/src/main/scala/scala/scalanative/runtime/libc.scala There is also any C code referenced in this area that needs to stay in `nativelib`. https://github.com/scala-native/scala-native/tree/master/nativelib/src/main/scala/scala/scalanative/runtime
2. There is the C/C++ standard library that is very well defined. https://en.cppreference.com/w/c/header
3. Then there is POSIX. @LeeTibbert, myself and others learned early on that we needed to follow the specification and certain guidelines that I think are mostly in our heads or written on other PRs. Latest is here. https://pubs.opengroup.org/onlinepubs/9699919799/

It can be very confusing as looking back and forth between the specs and the code and between the C and POSIX specs can get dizzying but it can be separated as far as I know.

If we move the code to the `posixlib` and the `clib` it should be easier the tease apart any code that is in the wrong place especially when the Windows port (hopefully someone will step forward) gets started in earnest.

**Comment by ekrich** [comment]
@sjrd and others. Do we want the `nativelib`, `posixlib`, and `clib` to omit the `exportJars := true` in `build.sbt`? 

We would need to `publishLocal` if we wanted to use the local `SNAPSHOT` and I would need to add some documentation in the Contributor section to that effect.

**Comment by LeeTibbert** [comment]
I concur that sorts based on experience gained and current knowledge & directions is the way to go.
In the medium range, as we work on posixlib stuff, we may want to see if it can be implemented as
a layer on top of clib.    At one point, I put cross-pointers in the clib & posixlib .rst files/tables to 
try and get a clean, or at least documented, description of what was expected in each library.

**Comment by ekrich** [comment]
@sjrd Do you think we could add this PR and a follow on for `clib` into the release so at least we get the associated native C code in the correct place?

The dependency chain currently `javalib` -> `posixlib` -> `clib` but technically it should be `javalib` to both. Then portions of the `javalib` once split would depend on `posixlib` or `clib` but not both.

The small dependency between `posixlib` and `clib` has been removed.

**Comment by ekrich** [comment]
I changed this to draft to address `clib` and any other changes needed.

**Comment by ekrich** [comment]
This now contains potential breaking changes to downstream applications.
* Some items are in clib and not in posix lib as per the specs
* Some posixlib items have been moved to the correct file so imports may need changing

Overall, we want to strive for correctness so I see no way around this as macro defs can only be in defined one place so deprecation is not possible. C is a flat name space.

**Comment by ekrich** [comment]
TODO:

[ ] Determine package for SocketHelpers, maybe `posix.util`?
[ ] Do we want `time` and Java specific portions of `Platform` in `runtime` or `javalib`?

File: nativelib/src/main/resources/scala-native/time_millis.c
Comment (LeeTibbert) [inline nativelib/src/main/resources/scala-native/time_millis.c:11]: I created issue #1948 to report that the error code for gettimeofday() is never checked.
I am not suggesting a change for this PR, just wanted to note that it had been detected.

File: clib/src/main/resources/scala-native/errno.c
Comment (LeeTibbert) [inline clib/src/main/resources/scala-native/errno.c:3]: float.c and math.c below seem to use the scalanative_libc_ prefix.  That seems to be a useful convention.  If not relevant here, to keep changes to a minimum, perhaps in a spin-off PR?

File: javalib/src/main/scala/java/io/File.scala
Comment (LeeTibbert) [inline javalib/src/main/scala/java/io/File.scala:44]: The change from fcntl.mumble to unistd.mumble in a number of places is a __long__ overdue correction.  Thank you!

File: nativelib/src/main/resources/scala-native/time_millis.c
Comment (LeeTibbert) [inline nativelib/src/main/resources/scala-native/time_millis.c:11]: To not distract the major intent of this PR, I opened Issue #1948 to capture the fact that the return value of gettimeofday() is never checked.  This is a bug from the code before this port, not this PR, so it should not hold up this PR.

File: posixlib/src/main/resources/scala-native/types.h
Comment (LeeTibbert) [inline posixlib/src/main/resources/scala-native/types.h:4]: OK, why?  I know the answer but that is surely the question that a new reader will ask.

File: posixlib/src/main/resources/scala-native/unistd.c
Comment (LeeTibbert) [inline posixlib/src/main/resources/scala-native/unistd.c:16]: Again, IMO, those three should be defined here, probably as calls to the clib version.
Independent duplication is also possible, if you/we want to avoid a dependency of posixlib on clib.

File: posixlib/src/main/resources/scala-native/unistd.c
Comment (LeeTibbert) [inline posixlib/src/main/resources/scala-native/unistd.c:4]: I did not check this file against the posix standard, but it looks better than what it
replaces.

File: posixlib/src/main/scala/scala/scalanative/posix/fcntl.scala
Comment (LeeTibbert) [inline posixlib/src/main/scala/scala/scalanative/posix/fcntl.scala:16]: Because I have been kicked by this file before, I _did_ check it against the posix specification.
I found it to be problematic, from before this PR.  I believe that an Issue should be created
and this file, in its new location, marked for scrutiny, probably by at least two people.

This file is known to be incomplete, especially the constants. I did not try to figure out the missing values.
I did, however, look at the example for fcntl().  By my reading, that can not be implemented in SN with
the current method declarations, or at least not cleanly.  fcntl() and, I believe, open() are defined in
posix as having a variable number of arguments.  That is difficult in SN.  If it is too difficult, then
the methods should be marked as non-posix but passably useful in the file.

File: posixlib/src/main/scala/scala/scalanative/posix/unistd.scala
Comment (LeeTibbert) [inline posixlib/src/main/scala/scala/scalanative/posix/unistd.scala:52]: The comment is good, but this is another case of you are posix or not.  posix defines null in unistd.h, I just looked.  At the least, the NULL here should be a reference
to the NULL in stddef (or visa versa).  That would make each file (more) posix compliant and reduce the amount of SN lore a user of SN needs to know.

File: posixlib/src/main/scala/scala/scalanative/posix/unistd.scala
Comment (LeeTibbert) [inline posixlib/src/main/scala/scala/scalanative/posix/unistd.scala:66]: Same song, next verse.

File: posixlib/src/main/scala/scala/scalanative/runtime/SocketHelpers.scala
Comment (LeeTibbert) [inline posixlib/src/main/scala/scala/scalanative/runtime/SocketHelpers.scala:26]: The change gets rid of a cast.  I spent more time that I admit trying to figure out which of the two uses of libc was relevant here, and in ratting through includes to figure it out. 

I believe that a special variant of libc and of memset() are being used. Withot
further info, I expected the libc variant (yes, that libc ;-) which one I leave as
an exercise for the Gentle Reader).
If I got it right, please consider `runtime.libc.memset`. I think that would give a 
better clue.

If I got it wrong, then I am clueless (could be! here and elsewhere) and could use
a better clue or more coffee.

File: posixlib/src/main/scala/scala/scalanative/runtime/SocketHelpers.scala
Comment (LeeTibbert) [inline posixlib/src/main/scala/scala/scalanative/runtime/SocketHelpers.scala:114]: Ibid.

**Comment by ekrich** [comment]
@LeeTibbert This PR is a moving code and aligning dependencies. The goal is to get the code into the correct position. You should make issues when you see stuff but any issue that gets fixed also creates a rebase/merge and we can never hope to get this PR to the finish line then. I suggest you make issues for each of the things above because they are out of scope for this PR.

I did the hard bulk work to get this closer to being correct but there is no hope of making it even close to perfect. Just moving the `posix` was really not an option so I moved the `clib` too and separated and fixed as many obvious things as possible.

It is also ripe to consider that if don't do some of this nasty structural stuff that could potentially break things for users now when few people using Scala Native then it will never get done.

File: clib/src/main/resources/scala-native/stdio.c
Comment (LeeTibbert) [inline clib/src/main/resources/scala-native/stdio.c:1]: I think this is a concrete case which captures some architectural disquiet I have been 
trying to put into sensible words.

From build.sbt it appears that the intent is that javalib & clib be independent and rely upon nativelib.  From the description of this PR, it appears that the intent is that nativelib be minimal/skinny/lightweight.  So far so good.

Now we come upon stdio.  It is defined in both posix and c.  To the best of my searching in a clone of the PR branch, there is a stdio.scala in clib (corresponding, I believe, to this C file) but _none_ in posix lib, where a posix developer would expect to find it.

There are at least four possibilities, none good, and I think it takes an architectural
decision in a README in each of posixlib and clib:

0) Let posixlib have holes as an explicit decision.

1) code duplication.  posixlib & clib maintain their own bugs and quirks. (Is my
    disregard for this solution showing? Oops!)

2) Posixlib calls into clib.  This violates one of the inferred goals of this PR.

3) Posixlib and clib call into a common lower layer. This gets up back to nativelib
    or such, which violates another intent.

**Comment by LeeTibbert** [comment]
re: This PR is a moving code and aligning dependencies

Understood, entirely, and agreed. PR such as this one do not
age well; however, they do age their creators, and probably not well.

 I was asked to go line by line and will spin off the Issues I identified, as 
this PR advances.

**Comment by LeeTibbert** [comment]
In summary, Weds 2020-10-14

As asked, I went over each of the 46 changed files line by line.

I also copied over Eric's posix-native branch, expanded it, and
looked around for my favorite files.  Things looked familiar enought
that I think I could figure out and get over the change for it is a more
understandable & rational separation.

To be through, but not hold up this PR, my next step is to try actually working in the environment
of this PR.

At the 100 meter level, I concur with the direction of this work and think it adds value.  It is probably 
not the last pass but it is IMO much better than the current setup.  As architectures are added the
layout will probably need to change, this PR untangles middle layer libraries and facilitates that.

At the 99 meter level, I think there is an architectural/design issue.  I discussed this in a comment to clib 
stdio.c.

At the 90 meter level, this is obviously a breaking change and so should probably
be done along with other breaking changes, such as 0.next. Otherwise 0.next
and 0.next.next will both be breaking releases.  Yes, there is a concern for 
tossing too many breaking changes on developers, being one myself.
Earlier, with some documentation guidance, is probably better than later, but not my call.

In the weeds, at the centimeter level I identified several concerns and suggested that they 
be spun off into separate PRs so that they do not delay or hold up this PR.  I can do that
after people have a chance to comment on my comments. Perhaps I am off base.
Most of these issues are in existing code, or in snippets taken from existing code.

Lastly, and probably most importantly, Thank you Eric for going boldly...

**Comment by ekrich** [comment]
Thanks for all your work doing this review. I would like to try and explain better the dependencies.

From the Scala Native point of view, see build.sbt. Depends on is represented by `->`
`javalib -> posixlib, clib`
This is expected and better than it was before with a transitory dependency on `nativelib` as shown next.
`posixlib, clib -> nativelib`
This is also as expected. Most of the code was in `nativelib`. Some of the remaining code is only related to `javalib` and should be moved there in another PR. Theoretically, even the optional jar/zip code should be in `javalib`.

From a C perspective.
`posixlib, clib -> libc.so/dylib/a`
C has a flat namespace so by definition. Anything defined in the C spec cannot be duplicated as defined in the POSIX spec. They use conditional macros in the header to add additional header declarations for POSIX above what is part of the C spec. There is only one header and implementation of `fopen` in `libc`.

From a Scala point of view if we look at `stdio.scala` we see `fopen` and `EOF` for example. These could be duplicated in both `posixlib` and `clib` because we have package namespaces or the user could use `clib` plus `posixlib` to get the full POSIX experience. The constants such as `EOF` could also be duplicated on the C side of things as long as they are `scalanative_clib_EOF` and `scalanative_posix_EOF`. I am pretty sure _not_ duplicating things could be harder for the user because they have to import from both the `clib` and the `posixlib` but easier on the Scala Native maintainer as there is one of each thing.

File: posixlib/src/main/scala/scala/scalanative/runtime/SocketHelpers.scala
Comment (ekrich) [reply posixlib/src/main/scala/scala/scalanative/runtime/SocketHelpers.scala:26]: This is using the special runtime `nativelib` `libc.memset` which is needed so that the runtime doesn't depend on `clib`. It uses this one so it doesn't depend on `clib`.

The `SocketHelpers.scala` really belongs in the portion of `javalib` that depends on `posixlib`. It is in the `runtime` package but this needs to be a temporary home until a future PR. It should be private to Scala Native so people should not be using it.

File: posixlib/src/main/resources/scala-native/unistd.c
Comment (ekrich) [reply posixlib/src/main/resources/scala-native/unistd.c:4]: Yes, this is better.

File: posixlib/src/main/resources/scala-native/types.h
Comment (ekrich) [reply posixlib/src/main/resources/scala-native/types.h:4]: I put that there so the reader could know it is used only.

File: posixlib/src/main/resources/scala-native/errno.c
Comment (ekrich) [reply posixlib/src/main/resources/scala-native/errno.c:3]: I did not duplicate but if they have to be one place then they should be in `clib`. For certain `clib` should be complete but how we handle the overlap with POSIX in unanswered at this time.

File: javalib/src/main/scala/java/io/File.scala
Comment (ekrich) [reply javalib/src/main/scala/java/io/File.scala:44]: I fixed everything I saw that was just wrong especially if I had to because of a dependency.

File: clib/src/main/resources/scala-native/errno.c
Comment (ekrich) [reply clib/src/main/resources/scala-native/errno.c:3]: I think the `scalanative` prefix is good and we may need a `libc` too but this obviously is not totally agreed upon territory.

File: nativelib/src/main/resources/scala-native/time_millis.c
Comment (ekrich) [reply nativelib/src/main/resources/scala-native/time_millis.c:11]: Good idea. An error in C is like an error of death.

File: posixlib/src/main/resources/scala-native/unistd.c
Comment (ekrich) [reply posixlib/src/main/resources/scala-native/unistd.c:16]: There is not specific dependency but I get the point - some of this discussed later.

File: posixlib/src/main/scala/scala/scalanative/posix/fcntl.scala
Comment (ekrich) [reply posixlib/src/main/scala/scala/scalanative/posix/fcntl.scala:16]: Maybe add an issue for this as I do not know.

File: posixlib/src/main/scala/scala/scalanative/posix/unistd.scala
Comment (ekrich) [reply posixlib/src/main/scala/scala/scalanative/posix/unistd.scala:52]: I was going to just add it but the type is weird on macOS so I avoided it.

NULL is there but extra info is in stddef - this is a note to future devs.

File: posixlib/src/main/scala/scala/scalanative/posix/unistd.scala
Comment (ekrich) [reply posixlib/src/main/scala/scala/scalanative/posix/unistd.scala:66]: OK

File: posixlib/src/main/scala/scala/scalanative/runtime/SocketHelpers.scala
Comment (ekrich) [reply posixlib/src/main/scala/scala/scalanative/runtime/SocketHelpers.scala:114]: See above.

File: clib/src/main/resources/scala-native/stdio.c
Comment (ekrich) [reply clib/src/main/resources/scala-native/stdio.c:1]: I think what you have above should definitely be captured in an issue and when determined in the developer docs. This is a tricky issue. I have thought about it some but don't know the answer or even the balance of the options.

File: posixlib/src/main/resources/scala-native/unistd.c
Comment (ekrich) [reply posixlib/src/main/resources/scala-native/unistd.c:16]: I looked at this and they were suppose to be in POSIX `stdio.scala` or `stdio.c` and there isn't one at all so this was the minimal fix that was actually correct.

File: posixlib/src/main/resources/scala-native/errno.c
Comment (ekrich) [reply posixlib/src/main/resources/scala-native/errno.c:3]: This too was just moving things around to get them in the logical place. The code basically has a C file that is named the same as the Scala file except the extension when macros are needed. We didn't always early on have a plan - stuff was thrown wherever.

File: clib/src/main/resources/scala-native/stdio.c
Comment (ekrich) [reply clib/src/main/resources/scala-native/stdio.c:1]: Even under its own admission POSIX extends the C standard - click on the [CX] to get some more info.
https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/stdio.h.html

**Review by sjrd — APPROVED** [review]

**Pull request merged**
